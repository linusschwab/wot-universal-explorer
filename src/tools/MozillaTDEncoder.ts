import {Thing} from "../models/thing";
import {Action, Event, Property} from "../models/interactions";
import {App} from "../app";


export class MozillaTDEncoder {

    public static encode(thing: Thing) {
        return JSON.stringify(thing, (key: string, value: Object) => {
            if (value instanceof Thing) {
                return MozillaTDEncoder.thing(value);
            }

            return value;
        }, 4);
    }

    private static thing(thing: Thing) {
        // TODO: href? Include in Thing?
        return {
            "name": thing.name,
            "type": thing.type,
            "href": "",
            "properties": MozillaTDEncoder.properties(thing.properties),
            "actions": MozillaTDEncoder.actions(thing.actions),
            "events": MozillaTDEncoder.events(thing.events)
        };
    }

    private static properties(properties: Property[]) {
        let obj: any = {};
        for (const property of properties) {
            obj[property.name] = {
                "type": property.schema.type,
                "description": property.description,
                "href": App.url + '/things' + property.url
            };

            if (!property.description) {
                delete obj.description;
            }
        }
        return obj;
    }

    private static actions(actions: Action[]) {
        let obj: any = {};
        for (const action of actions) {
            // TODO: Include href (as links object?)
            obj[action.name] = {
                "description": action.description,
            };
        }
        return obj;
    }

    private static events(events: Event[]) {
        let obj: any = {};
        for (const event of events) {
            // TODO: Include href (as links object?)
            obj[event.name] = {
                "description": event.description,
            };
        }
        return obj;
    }
}