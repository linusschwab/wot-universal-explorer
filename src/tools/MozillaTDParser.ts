import {MozillaThing} from "../models/thing";
import {Property} from "../models/interactions";
import {DataSchema} from "../models/schema";
import {MozillaHTTPLink} from "../models/links/MozillaHTTPLink";


export class MozillaTDParser {

    public static parse(td: string | object) {
        let obj;

        if (typeof td === 'string') {
            obj = JSON.parse(td);
        } else {
            obj = td;
        }

        const base = MozillaTDParser.parseBase(obj.href);
        let thing = new MozillaThing(obj.name, obj.type, base);
        thing.description = obj.description;

        for (let [name, pobj] of Object.entries(obj.properties)) {
            let property = MozillaTDParser.parseProperty(name, pobj);

            let link = new MozillaHTTPLink((<any>pobj).href, base);
            property.registerLink(link);

            thing.registerInteraction(property);
        }

        return thing;
    }

    private static parseBase(base: string) {
        if (base.includes('http://')) {
            return base;
        } else {
            return process.env.MOZ_BASE + base;
        }
    }

    private static parseProperty(name: string, pobj: any) {
        // TODO: include unit in data schema? (celsius, percentage, ...)
        let schema = new DataSchema(pobj.type, true, false);
        return new Property(name, schema, true, true);
    }
}
