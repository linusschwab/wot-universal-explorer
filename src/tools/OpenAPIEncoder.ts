import {Thing, ThingsManager} from "../models/thing";
import {Operation} from "../models/links";
import {DataSchema, InputSchema} from "../models/schema";
import {Action, InteractionPattern, Property, Event} from "../models/interaction";
import {isArray} from "util";


export class OpenAPIEncoder {

    public static encode(things: ThingsManager) {
        return JSON.stringify(things, (key: string, value: Object) => {
            if (value instanceof ThingsManager) {
                return {
                    "openapi": "3.0.0",
                    "info": this.info(),
                    "servers": this.servers(),
                    "paths": this.paths(value),
                }
            }
            if (value instanceof InteractionPattern) {
                return this.operations(value);
            }
            if (value instanceof Operation) {
                return this.operation(value);
            }

            return value;
        }, 4);
    }

    private static info() {
        return {
            "version": "1.0",
            "title": "Universal Explorer for the Web of Things",
            "description": "API to interact with things",
        };
    }

    private static servers() {
        // TODO: Remove hardcoded url, read from config
        return [
            {
                "url": "http://localhost:5000/things",
                "description": "Development server"
            }
        ]
    }

    private static paths(things: ThingsManager) {
        let obj: any = {};

        for (let interaction of things.getInteractions()) {
            obj[interaction.url] = interaction;
        }

        return obj
    }

    private static operations(interaction: InteractionPattern) {
        let obj: any = {};

        for (let operation of interaction.operations) {
            obj[operation.type] = operation;
        }

        return obj;
    }

    private static operation(o: Operation) {
        let operation: any = {
            "description": o.description,
            "responses": {
                "200": {
                    "description": "success"
                }
            },
            "tags": [
                o.interaction.thing.name
            ]
        };

        // Request body
        let requestSchema: DataSchema = this.requestSchema(o);
        if (requestSchema) {
            operation['requestBody'] = this.requestBody(
                o.interaction.toString(),
                'application/json',
                requestSchema
            );
        }

        // Response
        let responseSchema: DataSchema = this.responseSchema(o);
        if (responseSchema) {
            operation['responses']['200']['content'] = {
                ['application/json']: {
                    "schema": this.schema(responseSchema)
                }
            };
        }

        return operation;
    }

    private static requestSchema(o: Operation): DataSchema {
        let requestSchema: DataSchema = null;

        // Check if operation type allows request body
        if (o.type !== 'post' && o.type !== 'put') {
            return null;
        }

        // Transform interaction data schema
        if (o.interaction instanceof Property) {
            if (o.interaction.schema && o.interaction.schema.writable) {
                requestSchema = o.interaction.schema;
            }
        }
        else if (o.interaction instanceof Action) {
            requestSchema = o.interaction.inputSchema;
        }
        else if (o.interaction instanceof Event) {
            requestSchema = null;
        }

        return requestSchema
    }

    private static responseSchema(o: Operation): DataSchema {
        if (o.interaction instanceof Property) {
            return o.interaction.schema;
        }
        else if (o.interaction instanceof Action) {
            return o.interaction.outputSchema;
        }
        else if (o.interaction instanceof Event) {
            return o.interaction.schema;
        }
        else {
            return null;
        }
    }

    private static schema(data: DataSchema | DataSchema[]) {
        if (!isArray(data)) {
            data = [data];
        }

        // TODO: Check if same name multiple times
        let properties: any = {};
        let required: string[] = [];

        for (let d of data) {
            properties[d.type] = {
                "type": d.type
            };

            if (d instanceof InputSchema) {
                required.push(d.type);
            }
        }

        let schema: any = {
            "type": "object",
            "properties": properties
        };

        if (required.length > 0) {
            schema['required'] = required;
        }

        return schema;
    }

    private static requestBody(description: string, mediaType: string, data: DataSchema | DataSchema[]) {
        return {
            "description": description + " parameters",
            "required": "true",
            "content": {
                [mediaType]: {
                    "schema": this.schema(data)
                }
            }
        }
    }

}
