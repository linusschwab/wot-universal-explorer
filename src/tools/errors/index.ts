export {InteractionError} from "./InteractionError";
export {TimeoutError} from "./TimeoutError";
