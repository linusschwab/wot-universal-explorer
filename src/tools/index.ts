export {MozillaTDParser} from "./MozillaTDParser";
export {OpenAPIEncoder} from "./OpenAPIEncoder";
export {TDEncoder} from "./TDEncoder";
export {TDParser} from "./TDParser";
