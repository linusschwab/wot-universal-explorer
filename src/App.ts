import * as Koa from "koa";
import * as bodyParser from "koa-bodyparser";
import * as cors from "@koa/cors";
import * as WebSocket from "ws";
import * as fs from "fs";

import {Server} from "http";
import {ThingsManager} from "./models/thing";
import {MozillaTDParser, OpenAPIEncoder, TDParser} from "./tools";
import {ControllerManager} from "./controllers";


export class App {

    public koa: Koa;
    public server: Server;
    public wss: WebSocket.Server;

    public things: ThingsManager;
    public controllers: ControllerManager;

    public static port = Number(process.env.PORT) || 5000;
    public static url = 'http://localhost:' + App.port;
    public static instance: App;

    public static run() {
        App.instance = new App();

        App.instance.server = App.instance.koa.listen(App.port);
        App.instance.setupWebSocket();
        App.instance.importTD();

        console.log('Server listening on port ' + App.port);
    }

    constructor() {
        this.koa = new Koa();
        this.things = new ThingsManager();
        this.controllers = new ControllerManager(this.things);

        this.setupMiddleware();
        this.setupRoutes();
    }

    private setupMiddleware() {
        // Mount logger
        this.koa.use(async (ctx, next) => {
            const start = Date.now();
            await next();
            const ms = Date.now() - start;
            console.log(`${ctx.method} ${ctx.url} - ${ms}ms`);
        });

        // Mount body parser
        this.koa.use(bodyParser({strict: false}));

        // Enable CORS
        this.koa.use(cors());

        // Set defaults
        this.koa.use(async (ctx, next) => {
            ctx.set('Content-Type', 'application/json');
            await next();
        });
    }

    private setupRoutes() {
        this.koa.use(this.controllers.td.router.routes());
        this.koa.use(this.controllers.things.router.routes());
        // Must be defined last (generic route to serve static files)
        this.koa.use(this.controllers.index.router.routes());
    }

    private setupWebSocket() {
        this.wss = new WebSocket.Server({ port: 8080 });
        this.wss.on('connection', (ws, req) => this.controllers.things.wsConnection(ws, req));
    }

    private importTD() {
        // Parse td files to things
        const tdPath = '../public/td/';
        fs.readdirSync(tdPath).forEach(file => {
            let td = fs.readFileSync(tdPath + file, 'utf8');
            let thing = TDParser.parse(td);
            this.things.addThing(thing);
        });

        // Parse moz-td files to things
        const mozTDPath = '../public/td-moz/';
        fs.readdirSync(mozTDPath).forEach(file => {
            let td = fs.readFileSync(mozTDPath + file, 'utf8');
            let thing = MozillaTDParser.parse(td);
            this.things.addThing(thing);
        });

        // Generate openapi file from things
        let json = OpenAPIEncoder.encode(this.things);
        fs.writeFile('../public/swagger-ui/openapi.json', json, 'utf8', (err) => {
            if (err) {
                console.log(err.message);
            }
        });
    }
}
