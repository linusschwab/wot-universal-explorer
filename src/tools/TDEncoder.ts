import {Thing} from "../models/thing";
import {Action, InteractionPattern, Property, Event} from "../models/interactions";
import {DataSchema} from "../models/schema";
import {App} from "../App";


export class TDEncoder {

    public static encode(thing: Thing) {
        return JSON.stringify(thing, (key: string, value: Object) => {
            if (value instanceof Thing) {
                return TDEncoder.thing(value);
            }
            if (value instanceof InteractionPattern) {
                return TDEncoder.interaction(value);
            }

            return value;
        }, 4);
    }

    private static thing(thing: Thing) {
        return {
            "@context": ["https://w3c.github.io/wot/w3c-wot-td-context.jsonld",
                "https://w3c.github.io/wot/w3c-wot-common-context.jsonld"],
            "@type": [thing.type],
            "name": thing.name,
            "base": App.url + '/things',
            "interaction": thing.interactions
        };
    }

    // Interactions
    private static interaction(interaction: InteractionPattern) {
        if (interaction instanceof Property) {
            return TDEncoder.property(interaction);
        }
        else if (interaction instanceof Action) {
            return TDEncoder.action(interaction);
        }
        else if (interaction instanceof Event) {
            return TDEncoder.event(interaction);
        }

        // TODO: Throw error instead?
        return null;
    }

    private static property(property: Property) {
        return {
            "@type": ["Property"],
            "name": property.name,
            "schema": TDEncoder.schema(property.schema),
            "writable": property.writable,
            "observable": property.observable,
            "form": TDEncoder.form(property.url)
        }
    }

    private static action(action: Action) {
        return {
            "@type": ["Action"],
            "name": action.name,
            "inputSchema": TDEncoder.schema(action.inputSchema),
            "form": TDEncoder.form(action.url)
        }
    }

    private static event(event: Event) {
        return {
            "@type": ["Event"],
            "name": event.name,
            "schema": TDEncoder.schema(event.schema),
            "form": TDEncoder.form(event.url)
        }
    }

    private static schema(schema: DataSchema) {
        if (schema) {
            return {
                "type": schema.type
            }
        } else {
            return null;
        }
    }

    private static form(url: string) {
        return [
            {
                "href": url,
                "mediaType": "application/json"
            }
        ]
    }
}
