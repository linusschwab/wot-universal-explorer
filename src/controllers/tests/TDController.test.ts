import * as Koa from "koa";
import * as request from "supertest";
import {App} from "../../App";
import {OpenAPIEncoder} from "../../tools";


const tdSample = `{
  "@context": ["https://w3c.github.io/wot/w3c-wot-td-context.jsonld",
               "https://w3c.github.io/wot/w3c-wot-common-context.jsonld"],
  "@type": ["Lamp"],
  "name": "myMasterLED",
  "interaction": [
    {
      "@type": ["Action", "Toggle"],
      "name": "myMasterOnOff",
      "inputSchema": {
        "@type": "OnOff",
        "type": "boolean"
      },
      "form": [
        {
          "href" : "coap://www.example.com:5683/master",
          "mediaType": "application/json"
        },
        {
          "href" : "http://www.example.com:80/master",
          "mediaType": "application/json"
        }
      ]
    }
  ]
}`;

const mozTDSample = `{
    "name": "Virtual On/Off Color Light",
    "type": "onOffColorLight",
    "href": "/things/virtual-things-0",
    "properties": {
        "on": {
            "type": "boolean",
            "href": "/things/virtual-things-0/properties/on"
        },
        "color": {
            "type": "string",
            "href": "/things/virtual-things-0/properties/color"
        }
    },
    "actions": {},
    "events": {}
}`;

// Create mocks
console.log = jest.fn();
jest.mock('fs');
jest.mock('../../tools/OpenAPIEncoder');


let app: App;

beforeEach(() => {
    app = new App();
    jest.clearAllMocks();
});

describe('post td', () => {
    test('post empty td returns error', async () => {
        const response = await request(app.koa.callback())
            .post('/td/add')
            .send(null)
            .set('Content-Type', 'application/json');

        expect(response.status).toBe(400);
        expect(response.type).toBe('text/plain');
    });

    test('post td adds thing and encodes openapi file', async () => {
        const response = await request(app.koa.callback())
            .post('/td/add')
            .send(tdSample)
            .set('Content-Type', 'application/json');

        expect(response.status).toBe(200);
        expect(response.type).toBe('application/json');

        expect(app.things.things).toHaveLength(1);
        expect(OpenAPIEncoder.encode).toHaveBeenCalled();
    });

    test('post td then get td returns encoded td', async () => {
        await request(app.koa.callback())
            .post('/td/add')
            .send(tdSample)
            .set('Content-Type', 'application/json');

        const response = await request(app.koa.callback())
            .get('/td/mymasterled');

        expect(response.status).toBe(200);
        expect(response.type).toBe('application/json');
        expect(response.body).toMatchSnapshot();
    });
});

describe('post mozilla td', () => {
    test('post empty mozilla td returns error', async () => {
        const response = await request(app.koa.callback())
            .post('/td/moz/add')
            .send(null)
            .set('Content-Type', 'application/json');

        expect(response.status).toBe(400);
        expect(response.type).toBe('text/plain');
    });

    test('post mozilla td adds thing and encodes openapi file', async () => {
        const response = await request(app.koa.callback())
            .post('/td/moz/add')
            .send(mozTDSample)
            .set('Content-Type', 'application/json');

        expect(response.status).toBe(200);
        expect(response.type).toBe('application/json');

        expect(app.things.things).toHaveLength(1);
        expect(OpenAPIEncoder.encode).toHaveBeenCalled();
    });

    test('post mozilla td then get td returns encoded td', async () => {
        await request(app.koa.callback())
            .post('/td/moz/add')
            .send(mozTDSample)
            .set('Content-Type', 'application/json');

        const response = await request(app.koa.callback())
            .get('/td/moz/virtual-on-off-color-light');

        expect(response.status).toBe(200);
        expect(response.type).toBe('application/json');
        expect(response.body).toMatchSnapshot();
    });
});

describe('get td', () => {
    test('get td without adding a thing first returns error', async () => {
        const response = await request(app.koa.callback())
            .get('/td/counter');

        expect(response.status).toBe(404);
        expect(response.type).toBe('text/plain');
    });
});

describe('get mozilla td', () => {
    test('get mozilla td without adding a thing first returns error', async () => {
        const response = await request(app.koa.callback())
            .get('/td/moz/counter');

        expect(response.status).toBe(404);
        expect(response.type).toBe('text/plain');
    });
});
