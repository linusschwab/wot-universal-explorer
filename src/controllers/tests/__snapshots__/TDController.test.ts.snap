// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post mozilla td post mozilla td then get td returns encoded td 1`] = `
Object {
  "actions": Object {},
  "events": Object {},
  "href": "",
  "name": "Virtual On/Off Color Light",
  "properties": Object {
    "color": Object {
      "href": "http://localhost:5000/things/virtual-on-off-color-light/properties/color",
      "type": "string",
    },
    "on": Object {
      "href": "http://localhost:5000/things/virtual-on-off-color-light/properties/on",
      "type": "boolean",
    },
  },
  "type": "onOffColorLight",
}
`;

exports[`post td post td then get td returns encoded td 1`] = `
Object {
  "@context": Array [
    "https://w3c.github.io/wot/w3c-wot-td-context.jsonld",
    "https://w3c.github.io/wot/w3c-wot-common-context.jsonld",
  ],
  "@type": Array [
    "Lamp",
  ],
  "base": "http://localhost:5000/things",
  "interaction": Array [
    Object {
      "@type": Array [
        "Action",
      ],
      "form": Array [
        Object {
          "href": "/mymasterled/actions/mymasteronoff",
          "mediaType": "application/json",
        },
      ],
      "inputSchema": Object {
        "type": "boolean",
      },
      "name": "myMasterOnOff",
    },
  ],
  "name": "myMasterLED",
}
`;
